Lib/concurrent/futures/process.py

```diff
@@ -382,96 +364,139 @@ def shutdown_worker():
-                        new_pending_work_items = {}
-                        for work_id, work_item in pending_work_items.items():
-                            if not work_item.future.cancel():
-                                new_pending_work_items[work_id] = work_item
-
-                        pending_work_items = new_pending_work_items
-                        # Drain work_ids_queue since we no longer need to
-                        # add items to the call queue.
-                        while True:
-                            try:
-                                work_ids_queue.get_nowait()
-                            except queue.Empty:
-                                break
-
-                # Since no new work items can be added, it is safe to shutdown
-                # this thread if there are no pending work items.
-                if not pending_work_items:
-                    shutdown_worker()
-                    return
-            except Full:
-                # This is not a problem: we will eventually be woken up (in
-                # result_queue.get()) and be able to send a sentinel again.
-                pass
-        executor = None
+        return (_global_shutdown or executor is None
+                or executor._shutdown_thread)
+
+    def terminate_broken(self, cause):
+        # Terminate the executor because it is in broken state. The cause
+        # argument can be used to display more information on the error that
+        # leaded the executor to be broken.
+
+        # Mark the process pool broken so that submits fail right now.
+        executor = self.executor_reference()
+        if executor is not None:
+            executor._broken = ('A child process terminated '
+                                'abruptly, the process pool is not '
+                                'usable anymore')
+            executor._shutdown_thread = True
+            executor = None
+
+        # All pending tasks are to be marked failed with the following
+        # BrokenProcessPool error
+        bpe = BrokenProcessPool("A process in the process pool was "
+                                "terminated abruptly while the future was "
+                                "running or pending.")
+        if cause is not None:
+            bpe.__cause__ = _RemoteTraceback(
+                f"\n'''\n{''.join(cause)}'''")
+
+        # Mark pending tasks as failed.
+        for work_id, work_item in self.pending_work_items.items():
+            work_item.future.set_exception(bpe)
+            # Delete references to object. See issue16284
+            del work_item
+        self.pending_work_items.clear()
+
+        # Terminate remaining workers forcibly: the queues or their
+        # locks may be in a dirty state and block forever.
+        for p in self.processes.values():
+            p.terminate()
+
+        # clean up ressources
+        self.shutdown_worker()
+
+    def shutdown_executor_when_no_pending_tasks(self):
```

Comment:
Hmm, I'm really not a huge fan of the name `shutdown_executor_when_no_pending_tasks()`, it seems a bit long winded for a method name. I think just `shutdown_executor()` is fine as long as we clarify in a docstring that the resources are not finalized until there are no pending tasks remaining.